### PR TITLE
refactor: Use `merge_sort` instead of `qsort` for sorting.

### PR DESCRIPTION
**Pull request by iphydf**
This change is [Reviewable](https://reviewable.io/reviews/TokTok/c-toxcore/2638)

### CODE REVIEW

**Comment by JFreegman** [comment]
What are the advantages of merge sort over qsort in this instance?

**Comment by iphydf** [comment]
> What are the advantages of merge sort over qsort in this instance?

https://github.com/TokTok/c-toxcore/pull/2638/files#diff-421ac668286fc646ef81a78b41a2b05bb4c1dbf1f02f3ed69ced2929a8ba4e10R141-R145

I guess I can explain more in that comment if needed. This PR is part of my attempt to make toxcore C code rely less on struct layouts and low level system properties and be semantically simpler. I know C coders like to bit-twiddle everything, but I think it's worth trying to avoid that as much as possible. `qsort` operates with memcpy, lots of casting from and to `char*`, moving around in abstract memory with pointer arithmetic, etc. It's semantically complex and requires complicated abstract domains that understand memory layouts and their implications.

**Comment by iphydf** [comment]
As for why merge sort instead of any other algorithm: it's an easy one to implement and has deterministic time complexity. If we have a lot of mostly-sorted arrays, we should probably just sorted-insert elements rather than adding a bunch of unsorted elements and sorting afterwards.

File: toxcore/util_test.cc
Comment (Green-Sky) [inline toxcore/util_test.cc:1]: Why did you remove those tests. Also why include `<limits>` and not `<climits>`?

Comment (iphydf) [reply toxcore/util_test.cc:1]: I don't remember. Reverted.

Comment (iphydf) [reply toxcore/util_test.cc:1]: I remember now. They moved to crypto_core_test.cc.

**Comment by iphydf** [comment]
> with qsort gone, are we still using anything from `stdlib.h` ?

Only in group.c. I've removed it from the other 2.

**Review by Green-Sky — COMMENTED** [review]
and onion_announce.c ?

**Comment by iphydf** [comment]
> and onion_announce.c ?

Done.

**Review by Green-Sky — APPROVED** [review]
